micropub.php

```diff
@@ -49,389 +51,419 @@
 			}
+
+			kses_remove_filters();  // prevent sanitizing HTML tags in post_content
+			$args['ID'] = static::check_error( wp_insert_post( $args, true ) );
+			kses_init_filters();
+
+			static::default_file_handler( $args['ID'] );
+			return $args;
 		}
 
-		// replace
-		$replace = static::$input['replace'];
-		if ( $replace ) {
-			if ( ! is_array( $replace ) ) {
-				static::error( 400, 'replace must be an object' );
-			}
-			foreach ( static::mp_to_wp( array( 'properties' => $replace ) )
-					  as $name => $val ) {
-				$args[ $name ] = $val;
+		/*
+		 * Handle an update request.
+		 *
+		 * This really needs a db transaction! But we can't assume the underlying
+		 * MySQL db is InnoDB and supports transactions. :(
+		 */
+		private static function update() {
+			$post_id = url_to_postid( static::$input['url'] );
+			$args    = get_post( $post_id, ARRAY_A );
+			if ( ! $args ) {
+				static::error( 400, static::$input['url'] . ' not found' );
 			}
-		}
 
-		// delete
-		$delete = static::$input['delete'];
-		if ( $delete ) {
-			if ( is_assoc_array( $delete ) ) {
-				if ( array_diff( array_keys( $delete ), array( 'category', 'syndication' ) ) ) {
-					static::error( 400, 'can only delete individual values from category and syndication; other properties not supported' );
+			// add
+			$add = static::$input['add'];
+			if ( $add ) {
+				if ( ! is_array( $add ) ) {
+					static::error( 400, 'add must be an object' );
+				}
+				if ( array_diff( array_keys( $add ), array( 'category', 'syndication' ) ) ) {
+					static::error( 400, 'can only add to category and syndication; other properties not supported' );
 				}
-				$delete_args = static::mp_to_wp( array( 'properties' => $delete ) );
-				if ( $delete_args['tags_input'] ) {
-					$args['tags_input'] = array_diff( $args['tags_input'] ?: array(),
-													  $delete_args['tags_input']);
+				$add_args = static::mp_to_wp( array( 'properties' => $add ) );
+				if ( $add_args['tags_input'] ) {
+					// i tried wp_add_post_tags here, but it didn't work
+					$args['tags_input'] = array_merge(
+						$args['tags_input'] ?: array(),
+						$add_args['tags_input']
+					);
 				}
-				if ( $delete_args['post_category'] ) {
-					$args['post_category'] = array_diff($args['post_category'] ?: array(),
-														 $delete_args['post_category']);
+				if ( $add_args['post_category'] ) {
+					// i tried wp_set_post_categories here, but it didn't work
+					$args['post_category'] = array_merge(
+						$args['post_category'] ?: array(),
+						$add_args['post_category']
+					);
 				}
-			} elseif ( is_array( $delete ) ) {
-				foreach ( static::mp_to_wp( array( 'properties' => array_flip( $delete ) ) )
-						  as $name => $_ ) {
-					$args[ $name ] = NULL;
+			}
+
+			// replace
+			$replace = static::$input['replace'];
+			if ( $replace ) {
+				if ( ! is_array( $replace ) ) {
+					static::error( 400, 'replace must be an object' );
 				}
-				if ( in_array( 'category', $delete ) ) {
```

Comment:
this code change seems suspicious. was it unrelated and accidentally got lumped into this PR?

Reply:
It isn't deleted. It is just lower down on line 425.